File: forms/inpatient-clinical-note/front-end-with-svelte/src/lib/engine/news2.ts
// NEWS2 scoring, per the Royal College of Physicians 2017 report.
//
// See `doc/news2.md` for the full parameter tables and the escalation
// thresholds. This module derives the aggregate from the seven parameters; it
// never overwrites a total the clinician entered from the ward chart. The
// grader reports both, so a discrepancy between the chart and the parameters is
// visible rather than silently resolved.

import type { Observations } from './types';

export interface News2SubScores {
	respiratoryRate: number | null;
	oxygenSaturation: number | null;
	oxygenSupplement: number | null;
	systolicBloodPressure: number | null;
	pulse: number | null;
	consciousness: number | null;
	temperature: number | null;
}

export interface News2Derivation {
	/** Null when any of the seven parameters is missing. */
	total: number | null;
	subScores: News2SubScores;
	anyParameterScoresThree: boolean;
	complete: boolean;
}

/** True when a numeric parameter is usable. */
function num(v: number | null | undefined): v is number {
	return typeof v === 'number' && Number.isFinite(v);
}

/** Respiratory rate sub-score (breaths per minute). */
export function scoreRespiratoryRate(rr: number | null): number | null {
	if (!num(rr)) return null;
	if (rr <= 8) return 3;
	if (rr <= 11) return 1;
	if (rr <= 20) return 0;
	if (rr <= 24) return 2;
	return 3;
}

/**
 * Oxygen-saturation sub-score. Scale 2 applies only to patients with confirmed
 * hypercapnic respiratory failure and a prescribed target of 88-92%, and its
 * upper range scores differently depending on whether the patient is on oxygen.
 */
export function scoreOxygenSaturation(
	spo2: number | null,
	scale: string,
	onOxygen: boolean
): number | null {
	if (!num(spo2)) return null;
	if (scale === 'scale-2') {
		if (spo2 <= 83) return 3;
		if (spo2 <= 85) return 2;
		if (spo2 <= 87) return 1;
		if (spo2 <= 92) return 0;
		// Above the target range: scores only when the patient is on oxygen.
		if (!onOxygen) return 0;
		if (spo2 <= 94) return 1;
		if (spo2 <= 96) return 2;
		return 3;
	}
	if (spo2 <= 91) return 3;
	if (spo2 <= 93) return 2;
	if (spo2 <= 95) return 1;
	return 0;
}

/** Air-or-oxygen sub-score: 2 on supplemental oxygen, 0 on room air. */
export function scoreOxygenSupplement(oxygenDelivery: string): number | null {
	if (!oxygenDelivery) return null;
	return oxygenDelivery === 'air' ? 0 : 2;
}

/** Systolic blood-pressure sub-score (mmHg). */
export function scoreSystolicBloodPressure(sbp: number | null): number | null {
	if (!num(sbp)) return null;
	if (sbp <= 90) return 3;
	if (sbp <= 100) return 2;
	if (sbp <= 110) return 1;
	if (sbp <= 219) return 0;
	return 3;
}

/** Pulse sub-score (beats per minute). */
export function scorePulse(pulse: number | null): number | null {
	if (!num(pulse)) return null;
	if (pulse <= 40) return 3;
	if (pulse <= 50) return 1;
	if (pulse <= 90) return 0;
	if (pulse <= 110) return 1;
	if (pulse <= 130) return 2;
	return 3;
}

/** Consciousness sub-score: 0 when Alert, 3 for any of C, V, P, or U. */
export function scoreConsciousness(acvpu: string): number | null {
	if (!acvpu) return null;
	return acvpu === 'alert' ? 0 : 3;
}

/** Temperature sub-score (degrees Celsius). */
export function scoreTemperature(temp: number | null): number | null {
	if (!num(temp)) return null;
	if (temp <= 35.0) return 3;
	if (temp <= 36.0) return 1;
	if (temp <= 38.0) return 0;
	if (temp <= 39.0) return 1;
	return 2;
}

/** Per-parameter sub-scores. A missing parameter yields null for that row. */
export function scoreParameters(obs: Observations): News2SubScores {
	const onOxygen = !!obs.oxygenDelivery && obs.oxygenDelivery !== 'air';
	return {
		respiratoryRate: scoreRespiratoryRate(obs.respiratoryRate),
		oxygenSaturation: scoreOxygenSaturation(obs.oxygenSaturation, obs.spo2Scale, onOxygen),
		oxygenSupplement: scoreOxygenSupplement(obs.oxygenDelivery),
		systolicBloodPressure: scoreSystolicBloodPressure(obs.systolicBloodPressure),
		pulse: scorePulse(obs.pulseRate),
		consciousness: scoreConsciousness(obs.acvpu),
		temperature: scoreTemperature(obs.temperatureCelsius)
	};
}

/** Derive the NEWS2 aggregate from the seven parameters. */
export function deriveNews2(obs: Observations): News2Derivation {
	const subScores = scoreParameters(obs);
	const values = Object.values(subScores);
	const complete = values.every((v) => v !== null);
	const total = complete ? values.reduce((sum, v) => sum + (v as number), 0) : null;
	const anyParameterScoresThree = values.some((v) => v === 3);
	return { total, subScores, anyParameterScoresThree, complete };
}

export interface EffectiveNews2 {
	/** The total the acuity engine should use: entered wins over derived. */
	effective: number | null;
	entered: number | null;
	derived: number | null;
	anyParameterScoresThree: boolean;
}

/**
 * The NEWS2 total the acuity engine should use: the entered total when the
 * clinician recorded one, otherwise the derived total.
 */
export function effectiveNews2(obs: Observations): EffectiveNews2 {
	const d = deriveNews2(obs);
	const entered = num(obs.news2Total) ? obs.news2Total : null;
	return {
		effective: entered !== null ? entered : d.total,
		entered,
		derived: d.total,
		anyParameterScoresThree: d.anyParameterScoresThree
	};
}

/** Whether every one of the seven NEWS2 parameters has been recorded. */
export function hasFullObservationSet(obs: Observations): boolean {
	return deriveNews2(obs).complete;
}

File: forms/inpatient-clinical-note/back-end-with-loco/src/inpatient_clinical_note/engine/news2.rs
//! NEWS2 scoring, per the Royal College of Physicians 2017 report.
//!
//! See `doc/news2.md` for the parameter tables and the escalation thresholds.
//! This module derives the aggregate from the seven parameters; it never
//! overwrites a total the clinician entered from the ward chart. Both are
//! reported so a discrepancy is visible rather than silently resolved.

use super::types::Observations;

/// The seven NEWS2 sub-scores. `None` where the parameter was not recorded.
#[derive(Clone, Copy, Debug, Default)]
pub struct SubScores {
    /// Respiratory rate.
    pub respiratory_rate: Option<i32>,
    /// Oxygen saturation.
    pub oxygen_saturation: Option<i32>,
    /// Air or oxygen.
    pub oxygen_supplement: Option<i32>,
    /// Systolic blood pressure.
    pub systolic_blood_pressure: Option<i32>,
    /// Pulse.
    pub pulse: Option<i32>,
    /// Consciousness.
    pub consciousness: Option<i32>,
    /// Temperature.
    pub temperature: Option<i32>,
}

impl SubScores {
    fn all(self) -> [Option<i32>; 7] {
        [
            self.respiratory_rate,
            self.oxygen_saturation,
            self.oxygen_supplement,
            self.systolic_blood_pressure,
            self.pulse,
            self.consciousness,
            self.temperature,
        ]
    }
}

/// The result of deriving NEWS2 from the seven parameters.
#[derive(Clone, Copy, Debug, Default)]
pub struct Derivation {
    /// `None` when any of the seven parameters is missing.
    pub total: Option<i32>,
    /// The per-parameter sub-scores.
    pub sub_scores: SubScores,
    /// Whether any single parameter scores 3.
    pub any_parameter_scores_three: bool,
    /// Whether all seven parameters were recorded.
    pub complete: bool,
}

/// Respiratory-rate sub-score.
#[must_use]
pub fn score_respiratory_rate(rr: Option<i32>) -> Option<i32> {
    let rr = rr?;
    Some(match rr {
        i32::MIN..=8 => 3,
        9..=11 => 1,
        12..=20 => 0,
        21..=24 => 2,
        _ => 3,
    })
}

/// Oxygen-saturation sub-score. Scale 2 applies only to a prescribed target of
/// 88–92% in confirmed hypercapnic respiratory failure, and above that range it
/// scores only when the patient is on oxygen.
#[must_use]
pub fn score_oxygen_saturation(spo2: Option<i32>, scale: &str, on_oxygen: bool) -> Option<i32> {
    let spo2 = spo2?;
    if scale == "scale-2" {
        return Some(match spo2 {
            i32::MIN..=83 => 3,
            84..=85 => 2,
            86..=87 => 1,
            88..=92 => 0,
            93..=94 if on_oxygen => 1,
            95..=96 if on_oxygen => 2,
            _ if on_oxygen => 3,
            _ => 0,
        });
    }
    Some(match spo2 {
        i32::MIN..=91 => 3,
        92..=93 => 2,
        94..=95 => 1,
        _ => 0,
    })
}

/// Air-or-oxygen sub-score: 2 on supplemental oxygen, 0 on room air.
#[must_use]
pub fn score_oxygen_supplement(oxygen_delivery: &str) -> Option<i32> {
    if oxygen_delivery.is_empty() {
        return None;
    }
    Some(i32::from(oxygen_delivery != "air") * 2)
}

/// Systolic blood-pressure sub-score.
#[must_use]
pub fn score_systolic_blood_pressure(sbp: Option<i32>) -> Option<i32> {
    let sbp = sbp?;
    Some(match sbp {
        i32::MIN..=90 => 3,
        91..=100 => 2,
        101..=110 => 1,
        111..=219 => 0,
        _ => 3,
    })
}

/// Pulse sub-score.
#[must_use]
pub fn score_pulse(pulse: Option<i32>) -> Option<i32> {
    let pulse = pulse?;
    Some(match pulse {
        i32::MIN..=40 => 3,
        41..=50 => 1,
        51..=90 => 0,
        91..=110 => 1,
        111..=130 => 2,
        _ => 3,
    })
}

/// Consciousness sub-score: 0 when Alert, 3 for any of C, V, P, or U.
#[must_use]
pub fn score_consciousness(acvpu: &str) -> Option<i32> {
    if acvpu.is_empty() {
        return None;
    }
    Some(i32::from(acvpu != "alert") * 3)
}

/// Temperature sub-score.
#[must_use]
pub fn score_temperature(temp: Option<f64>) -> Option<i32> {
    let temp = temp?;
    Some(if temp <= 35.0 {
        3
    } else if temp <= 36.0 {
        1
    } else if temp <= 38.0 {
        0
    } else if temp <= 39.0 {
        1
    } else {
        2
    })
}

/// Derive the NEWS2 aggregate from the seven parameters.
#[must_use]
pub fn derive_news2(obs: &Observations) -> Derivation {
    let on_oxygen = !obs.oxygen_delivery.is_empty() && obs.oxygen_delivery != "air";
    let sub_scores = SubScores {
        respiratory_rate: score_respiratory_rate(obs.respiratory_rate),
        oxygen_saturation: score_oxygen_saturation(
            obs.oxygen_saturation,
            &obs.spo2_scale,
            on_oxygen,
        ),
        oxygen_supplement: score_oxygen_supplement(&obs.oxygen_delivery),
        systolic_blood_pressure: score_systolic_blood_pressure(obs.systolic_blood_pressure),
        pulse: score_pulse(obs.pulse_rate),
        consciousness: score_consciousness(&obs.acvpu),
        temperature: score_temperature(obs.temperature_celsius),
    };
    let all = sub_scores.all();
    let complete = all.iter().all(Option::is_some);
    let total = if complete {
        Some(all.iter().map(|v| v.unwrap_or(0)).sum())
    } else {
        None
    };
    let any_parameter_scores_three = all.contains(&Some(3));
    Derivation {
        total,
        sub_scores,
        any_parameter_scores_three,
        complete,
    }
}

/// The NEWS2 total the acuity engine should use, plus what it was derived from.
#[derive(Clone, Copy, Debug, Default)]
pub struct Effective {
    /// Entered wins over derived.
    pub effective: Option<i32>,
    /// The total entered from the ward chart.
    pub entered: Option<i32>,
    /// The total derived from the seven parameters.
    pub derived: Option<i32>,
    /// Whether any single parameter scores 3.
    pub any_parameter_scores_three: bool,
}

/// Resolve the NEWS2 total: an entered total always wins over a derived one.
#[must_use]
pub fn effective_news2(obs: &Observations) -> Effective {
    let d = derive_news2(obs);
    Effective {
        effective: obs.news2_total.or(d.total),
        entered: obs.news2_total,
        derived: d.total,
        any_parameter_scores_three: d.any_parameter_scores_three,
    }
}

/// Whether every one of the seven NEWS2 parameters has been recorded.
#[must_use]
pub fn has_full_observation_set(obs: &Observations) -> bool {
    derive_news2(obs).complete
}
